mypyc/genops.py

```diff
@@ -307,23 +307,59 @@ def visit_import_all(self, node: ImportAll) -> Value:
         return INVALID_VALUE
 
     def gen_func_def(self, fdef: FuncDef, class_name: Optional[str] = None) -> FuncIR:
-        self.enter()
+        self.environment.add_func(fdef)
 
+        # If there is more than one environment in the environment stack, then we are visiting a
+        # non-global function, and set a flag to note this.
+        is_nested = len(self.environments) > 1
+
+        self.enter(fdef.name())
+
+        if is_nested:
+            # If this is a nested function, then add a 'self' field to the environment, since we
+            # will be instantiating the function as a method of a new class representing that
+            # original function.
+            self.environment.add_local(Var('self'), object_rprimitive, is_arg=True)
         for arg in fdef.arguments:
             assert arg.variable.type, "Function argument missing type"
             self.environment.add_local(arg.variable, self.type_to_rtype(arg.variable.type),
                                        is_arg=True)
-        self.ret_type = self.convert_return_type(fdef)
+        self.environment.ret_type = self.convert_return_type(fdef)
+
         fdef.body.accept(self)
 
-        if is_none_rprimitive(self.ret_type) or is_object_rprimitive(self.ret_type):
+        if (is_none_rprimitive(self.environment.ret_type) or
+                is_object_rprimitive(self.environment.ret_type)):
             self.add_implicit_return()
         else:
             self.add_implicit_unreachable()
 
         blocks, env = self.leave()
         args = self.convert_args(fdef)
-        return FuncIR(fdef.name(), class_name, self.module_name, args, self.ret_type, blocks, env)
+
+        if is_nested:
+            # If this is a nested function, create a callable class to represent it, and change the
+            # function to instead be the '__call__' method of the newly created class. The class
+            # name is generated using the names of the functions that enclose this. A 'self'
+            # argument is added to the function which gets changed into a class method.
+            namespace = self.generate_function_namespace()
+            class_name = '{}_{}_obj'.format(fdef.name(), namespace)
+            args.insert(0, RuntimeArg('self', object_rprimitive))
+            func_ir = FuncIR('__call__', class_name, self.module_name, args, env.ret_type, blocks,
+                             env)
+            class_ir = ClassIR(class_name, self.module_name)
+            class_ir.methods.append(func_ir)
+            self.classes.append(class_ir)
+
+            # Whenever a function definition is encountered, instantiate the callable class
+            # representing the function, and load it into a register in the current environment.
+            temp_reg = self.load_function_class(fdef, namespace)
+            func_reg = self.environment.add_local(fdef, object_rprimitive)
+            self.add(Assign(func_reg, temp_reg))
```

Comment:
This is a lot of nested function specific stuff here, and nested functions are not the common case. Can it be pulled out into another method nicely?